Fixes to minor do_delete_user bugs

mateuszmandera opened the pull request:
https://chat.zulip.org/#narrow/stream/31-production-help/topic/delete_user for context on the first bag with `is_active=True` on the dummy user. The second bug is explained in the commit message.

Do we want to write a migration to fix older dummy users to switch their `is_active=False`?

timabbott commented:
Made one comment; otherwise lgtm aside from the test failure (which hopefully is the result of the test being wrong)

timabbott commented on zerver/lib/actions.py:
```diff
@@ -1200,6 +1200,7 @@ def do_delete_user(user_profile: UserProfile) -> None:
         full_name=f"Deleted User {user_id}",
+        active=False,
         is_mirror_dummy=True,
     )
```
And while it's at it, I guess this should fix the email address too?  Sigh.

mateuszmandera commented:
> Made one comment; otherwise lgtm aside from the test failure (which hopefully is the result of the test being wrong)

@timabbott  Yeah, that was just the 2nd commit not adjusting a test.

Added a migration to fix both bugs for old objects.

mateuszmandera commented:
@timabbott 

1. Added RealmAuditLogs for the user deletion, with a backfill migration
2. I added `date_joined` preservation - I'm wondering whether we actually want to preserve the other details? I think it's just not clear to me what the reason for it would be - if a user is to be "removed" with information being preserved, there's user deactivation for that, while I was thinking of "delete user" as a hard-delete where generally everything gets purged :thinking:
3. Sadly we can't backfill `date_joined` for older dummy users I think. The issue is that `modified_user` (and `acting_user` and `modified_stream` which mean analogical potential issues) is a ForeignKey with `on_delete=CASCADE` - so these entries have been getting deleted during the `user_profile.delete()` call. I think we probably want to consider this a bug, because we'd like to preserve `RealmAuditLog`, no matter what happens with the modified/acting objects? This sounds more tricky to resolve, so I'm not sure if it's best to work on it in this PR or just open a follow-up issues.

I've added `release goal` to this PR so it doesn't get missed, since it looks likely to me that we'd want these bugs here addressed for the release (this was brought up by a user in https://chat.zulip.org/#narrow/stream/49-development-help/topic/Deleting.20users ) - feel free to drop the label if I'm wrong in my assumption :smiley:

timabbott commented:
Merged, thanks @mateuszmandera!  I agree we want this in 5.0; if it didn't require a migration I'd probably have aimed for backporting to 4.x, but I think it's not worth it given those complications.

I'll start a conversation about the remaining follow-up questions.
